src/lib_ccx/ccx_decoders_708_output.c

```diff
@@ -0,0 +1,316 @@
+
+	if (_dtvcc_is_caption_empty(decoder))
+		return;
+
+	if (decoder->tv->time_ms_show + decoder->subs_delay < 0) // Drop screens that because of subs_delay start too early
+		return;
+
+	decoder->cc_count++;
+
+	char *buf = (char *) encoder->buffer;
+
+	for (int i = 0; i < DTVCC_SCREENGRID_ROWS; i++)
+	{
+		if (!_dtvcc_is_row_empty(decoder, i))
+		{
+			buf[0] = 0;
+
+			if (encoder->transcript_settings->showStartTime)
+				mstime_sprintf(decoder->tv->time_ms_show + decoder->subs_delay,
+							   "%02u:%02u:%02u,%03u|", buf + strlen(buf));
+
+			if (encoder->transcript_settings->showEndTime)
+				mstime_sprintf(decoder->tv->time_ms_hide + decoder->subs_delay,
+							   "%02u:%02u:%02u,%03u|", buf + strlen(buf));
+
+			if (encoder->transcript_settings->showCC)
+				sprintf(buf + strlen(buf), "CC1|"); //always CC1 because CEA-708 is field-independent
+
+			if (encoder->transcript_settings->showMode)
+				sprintf(buf + strlen(buf), "POP|"); //TODO caption mode(pop, rollup, etc.)
+
+			if (strlen(buf))
+				write(decoder->fh, buf, strlen(buf));
+
+			_dtvcc_write_row(decoder, i);
+			write(decoder->fh, encoder->encoded_crlf, encoder->encoded_crlf_length);
+		}
+	}
+	write(decoder->fh, encoder->encoded_crlf, encoder->encoded_crlf_length);
+}
+
+void _dtvcc_write_sami_header(dtvcc_service_decoder *decoder)
+{
+	struct encoder_ctx *encoder = ccx_dtvcc_ctx.encoder;
+	char *buf = (char *) encoder->buffer;
+	memset(buf, 0, INITIAL_ENC_BUFFER_CAPACITY);
+	size_t buf_len = 0;
+
+	buf_len += sprintf(buf + buf_len, "<sami>%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "<head>%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "<style type=\"text/css\">%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "<!--%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len,
+					   "p {margin-left: 16pt; margin-right: 16pt; margin-bottom: 16pt; margin-top: 16pt;%s",
+					   encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len,
+					   "text-align: center; font-size: 18pt; font-family: arial; font-weight: bold; color: #f0f0f0;}%s",
+					   encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, ".unknowncc {Name:Unknown; lang:en-US; SAMIType:CC;}%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "-->%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "</style>%s", encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "</head>%s%s", encoder->encoded_crlf, encoder->encoded_crlf);
+	buf_len += sprintf(buf + buf_len, "<body>%s", encoder->encoded_crlf);
+
+	write(decoder->fh, buf, buf_len);
+}
+
+void _dtvcc_write_sami_footer(dtvcc_service_decoder *decoder)
+{
+	struct encoder_ctx *encoder = ccx_dtvcc_ctx.encoder;
+	char *buf = (char *) encoder->buffer;
+	sprintf(buf, "</body></sami>");
+	write(decoder->fh, buf, strlen(buf));
+	write(decoder->fh, encoder->encoded_crlf, encoder->encoded_crlf_length);
+}
+
+void ccx_dtvcc_write_sami(dtvcc_service_decoder *decoder)
```

Comment:
Is it possible to use existing encoder ctx and you might have to return cc_subtitle struct after filling it,